index.ts

```diff
@@ -1,176 +1,191 @@
   if (ts.isVariableDeclaration(node)) {
-    cb(node)
+    cb(node);
   } else {
-    ts.forEachChild(node, n => forAllVarDecls(n, cb))
+    ts.forEachChild(node, (n) => forAllVarDecls(n, cb));
   }
 }
 
-function findFirstJsxOpeningLikeElementWithName(
+function findJsxOpeningLikeElementsWithName(
   node: ts.SourceFile,
-  tagName: string,
-  dm?: boolean
+  tagName: ElementName,
 ) {
-  var res: LooseObject[] = [];
-  find(node);
-
-  function find(node: ts.Node | ts.SourceFile): undefined {
-    if (!node) {
-      return undefined;
+  const messages: ts.JsxOpeningLikeElement[] = [];
+  function findJsxElement(n: ts.Node): undefined {
+    // Is this a JsxElement with an identifier name?
+    if (
+      ts.isJsxOpeningLikeElement(n) &&
+      ts.isIdentifier(n.tagName)
+    ) {
+      // Does the tag name match what we're looking for?
+      const childTagName = n.tagName;
+      if (childTagName.text === tagName) {
+        // node is a JsxOpeningLikeElement
+        messages.push(n);
+      }
     }
-    if (dm && ts.isSourceFile(node)) {
-      // getNamedDeclarations is not currently public
-      forAllVarDecls(node, (el: ts.Declaration) => {
-        if (isDefineMessages(el, tagName)) {
-          if (
-            ts.isCallExpression(el.initializer) &&
-            el.initializer.arguments.length
-          ) {
-            var nodeProps = el.initializer.arguments[0];
-            var props = findProps(nodeProps, tagName);
-            // props is an array of LooseObject
-            res = res.concat(props);
-          }
-        }
-      })
-    } else {
-      // Is this a JsxElement with an identifier name?
+    return ts.forEachChild(n, findJsxElement);
+  }
+  findJsxElement(node);
+  return messages;
+}
+
+function findMethodCallsWithName(
+  sourceFile: ts.SourceFile,
+  methodName: MethodName,
+  extractMessages: MessageExtracter,
+) {
+  let messages: Message[] = [];
+  // getNamedDeclarations is not currently public
+  forAllVarDecls(sourceFile, (decl: ts.Declaration) => {
+    if (isMethodCall(decl, methodName)) {
       if (
-        ts.isJsxOpeningLikeElement(node) &&
-        ts.isIdentifier(node.tagName)
+        ts.isCallExpression(decl.initializer) &&
+        decl.initializer.arguments.length
       ) {
-        // Does the tag name match what we're looking for?
-        const childTagName = node.tagName;
-        if (childTagName.text === tagName) {
-          // node is a JsxOpeningLikeElement
-          res.push(node);
-        }
+        const nodeProps = decl.initializer.arguments[0];
+        const declMessages = extractMessagesForNode(nodeProps, extractMessages);
+        messages = messages.concat(declMessages);
       }
     }
-
-    return ts.forEachChild(node, find);
-  }
-
-  return res;
+  });
+  return messages;
 }
+
 /**
  * Parse tsx files
  *
  * @export
  * @param {string} contents
  * @returns {array}
  */
+// TODO perhaps we should expose the Message interface
```

Comment:
@bang88 How would you feel about exposing the Message interface as the return type for the (publicly exported) `main` method (well, an array of them to be exact)? (As opposed to an array of plain object `{}`.)

It shouldn't break any consumers, because I believe it's a narrower type, but I'm still very new to TypeScript, so I could be wrong about that.  In practice the consumers seem to be plain JavaScript npm scripts (unless they use ts-node), so they probably don't care about types. Still, I like types as documentation.

Reply:
I like the types too. If you prefer or think is useful for you please do it.

Reply:
Thinking about it a bit more, I think it would be better if the interface looked more like this:

```ts
interface MessageDescriptor {
  id: string
  defaultMessage: string
  description?: string
  [key: string]: string; // for any other properties
}
```
Otherwise exposing it won't really add much value.